contract/models/contract.py

```diff
@@ -595,19 +599,44 @@ def _recurring_create_invoice(self, date_ref=False):
         return moves
 
     @api.model
-    def cron_recurring_create_invoice(self, date_ref=None):
+    def _cron_recurring_create(self, date_ref=False, create_type="invoice"):
+        """
+        The cron function in order to create recurrent documents
+        from contracts.
+        """
+        _recurring_create_func = f"_recurring_create_{create_type}"
+        if not hasattr(self, _recurring_create_func):
+            _logger.info(
+                _(
+                    "No function to create %s documents automatically is "
+                    "declared in contract.contract model. Passing."
+                ),
+                create_type,
+            )
+            return False
         if not date_ref:
             date_ref = fields.Date.context_today(self)
         domain = self._get_contracts_to_invoice_domain(date_ref)
-        invoices = self.env["account.move"]
+        domain = expression.AND(
+            [
+                domain,
+                [("generation_type", "=", create_type)],
+            ]
+        )
+        contracts = self.search(domain)
+        companies = set(contracts.mapped("company_id"))
         # Invoice by companies, so assignation emails get correct context
-        companies_to_invoice = self.read_group(domain, ["company_id"], ["company_id"])
-        for row in companies_to_invoice:
-            contracts_to_invoice = self.search(row["__domain"]).with_context(
-                allowed_company_ids=[row["company_id"][0]]
-            )
-            invoices |= contracts_to_invoice._recurring_create_invoice(date_ref)
-        return invoices
+        # Use ids instead of recordset for perfomances reasons
+        for company in companies:
+            contracts_to_invoice = contracts.filtered(
+                lambda c: c.company_id == company
+            ).with_company(company)
+            getattr(contracts_to_invoice, _recurring_create_func)(date_ref)
+        return True
+
+    @api.model
+    def cron_recurring_create_invoice(self, date_ref=None):
+        return self._cron_recurring_create(date_ref)
```

Comment:
I would suggest the explicit form for avoiding misunderstandings:

```suggestion
        return self._cron_recurring_create(date_ref, create_type="invoice")
```